src/server/pfs/server/commit_set.go

```diff
@@ -0,0 +1,455 @@
+			var repoCommit *pfs.Commit
+			var err error
+			if c, ok := newRepoCommits[pfsdb.RepoKey(bi.Branch.Repo)]; !ok {
+				repoCommit, err = d.makeEmptyCommit(txnCtx, bi.Branch, bi.DirectProvenance, nil)
+				if err != nil {
+					return err
+				}
+				newRepoCommits[pfsdb.RepoKey(bi.Branch.Repo)] = repoCommit
+			} else {
+				repoCommit = c
+			}
+			bi.Head = repoCommit
+			return nil
+		}); err != nil {
+			return errors.Wrapf(err, "error updating branch %s", bi.Branch)
+		}
+	}
+	// update parent/child relationships on commits where necessary.
+	// for each deleted commit, update its parent's children and childrens' parents.
+	parentsToNewChildren := make(map[*pfs.Commit]map[*pfs.Commit]struct{})
+	parentsToDeleteChildren := make(map[*pfs.Commit]map[*pfs.Commit]struct{})
+	childrenToNewParent := make(map[*pfs.Commit]*pfs.Commit)
+	for _, ci := range commitInfos {
+		parent := oldestAncestor(ci, deleteCommits)
+		newDescendants := traverseToEdges(ci, deleteCommits, func(commitInfo *pfs.CommitInfo) []*pfs.Commit {
+			return commitInfo.ChildCommits
+		})
+		for _, descendant := range newDescendants {
+			childrenToNewParent[descendant] = parent
+		}
+		if parent != nil {
+			deleteChildren, ok := parentsToDeleteChildren[parent]
+			if !ok {
+				deleteChildren = make(map[*pfs.Commit]struct{})
+				parentsToDeleteChildren[parent] = deleteChildren
+			}
+			deleteChildren[ci.Commit] = struct{}{}
+			newChildren, ok := parentsToNewChildren[parent]
+			if !ok {
+				newChildren = make(map[*pfs.Commit]struct{})
+				parentsToNewChildren[parent] = newChildren
+			}
+			for _, child := range newDescendants {
+				newChildren[child] = struct{}{}
+			}
+		}
+	}
+	for child, parent := range childrenToNewParent {
+		childInfo := &pfs.CommitInfo{}
+		if err := d.commits.ReadWrite(txnCtx.SqlTx).Update(child, childInfo, func() error {
+			childInfo.ParentCommit = parent
+			return nil
+		}); err != nil {
+			return errors.Wrapf(err, "error updating parent/child pointers")
+		}
+	}
+	for parent := range parentsToNewChildren {
+		parentInfo := &pfs.CommitInfo{}
+		if err := d.commits.ReadWrite(txnCtx.SqlTx).Update(parent, parentInfo, func() error {
+			childrenSet := make(map[string]*pfs.Commit)
+			for _, fc := range parentInfo.ChildCommits {
+				childrenSet[pfsdb.CommitKey(fc)] = fc
+			}
+			for newChild := range parentsToNewChildren[parent] {
+				childrenSet[pfsdb.CommitKey(newChild)] = newChild
+			}
+			for deleteChild := range parentsToDeleteChildren[parent] {
+				delete(childrenSet, pfsdb.CommitKey(deleteChild))
+			}
+			parentInfo.ChildCommits = make([]*pfs.Commit, 0)
+			for _, c := range childrenSet {
+				parentInfo.ChildCommits = append(parentInfo.ChildCommits, c)
+			}
+			return nil
+		}); err != nil {
+			return errors.Wrapf(err, "error updating parent/child pointers")
+		}
+	}
+	return nil
+}
+
+func (d *driver) checkSubvenantCommitSets(txnCtx *txncontext.TransactionContext, commitsets []*pfs.CommitSet) error {
```

Comment:
I think this becomes much simpler if the API is based on single commit sets rather than batches.